src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1911 @@
+        bool newest, available, installed, obsolete, protected;
+        JsonVariant *contents;
+        JsonDispatch dispatch_table[] = {
+                { "version",       JSON_VARIANT_STRING,  json_dispatch_const_string,  PTR_TO_SIZE(&version),   0               },
+                { "newest",        JSON_VARIANT_BOOLEAN, json_dispatch_boolean,       PTR_TO_SIZE(&newest),    0               },
+                { "available",     JSON_VARIANT_BOOLEAN, json_dispatch_boolean,       PTR_TO_SIZE(&available), 0               },
+                { "installed",     JSON_VARIANT_BOOLEAN, json_dispatch_boolean,       PTR_TO_SIZE(&installed), 0               },
+                { "obsolete",      JSON_VARIANT_BOOLEAN, json_dispatch_boolean,       PTR_TO_SIZE(&obsolete),  0               },
+                { "protected",     JSON_VARIANT_BOOLEAN, json_dispatch_boolean,       PTR_TO_SIZE(&protected), 0               },
+                { "changelog-url", JSON_VARIANT_STRING,  json_dispatch_const_string,  PTR_TO_SIZE(&changelog), JSON_PERMISSIVE },
+                { "contents",      JSON_VARIANT_ARRAY,   json_dispatch_variant_noref, PTR_TO_SIZE(&contents),  0               },
+                {}
+        };
+        int r;
+
+        assert(json);
+
+        r = json_dispatch(json, dispatch_table, JSON_MANDATORY, NULL);
+        if (r < 0)
+                return r;
+
+        assert(version);
+        assert(contents);
+
+        r = sd_bus_message_new_method_return(msg, &reply);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_open_container(reply, 'a', "{sv}");
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_append(reply, "{sv}", "version", "s", version);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_append(reply, "{sv}", "newest", "b", newest);
+        if (r < 0)
+                return r;
+        r = sd_bus_message_append(reply, "{sv}", "available", "b", available);
+        if (r < 0)
+                return r;
+        r = sd_bus_message_append(reply, "{sv}", "installed", "b", installed);
+        if (r < 0)
+                return r;
+        r = sd_bus_message_append(reply, "{sv}", "obsolete", "b", obsolete);
+        if (r < 0)
+                return r;
+        r = sd_bus_message_append(reply, "{sv}", "protected", "b", protected);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_append(reply, "{sv}", "changelog-url", "s", changelog);
+        if (r < 0)
+                return r;
+
+        /* "contents" is just a serialized table that we want to present from updatectl.
+         * So, let's make this a "private" key and just send across raw JSON. */
+        r = json_variant_format(contents, 0, &contents_str);
+        if (r < 0)
+                return r;
+        r = sd_bus_message_append(reply, "{sv}", "_contents", "s", contents_str);
+        if (r < 0)
+                return r;
```

Comment:
i don't grok what this is about? can you explain? feels like cheating?

Reply:
100% is cheating 

The contents table is just way too complicated to serialize and then deserializse for, frankly, no reason. It's just there for display purposes in updatectl. I saw no proper documentable use-case to implement this as a documented, public API. So hence this hack to make life easier

Reply:
so i guess the lesson here is "varlink FTW!"?

Reply:
so, we started adding Describe() methods to some of the other daemons, specifically hostnamed and networkd. In both cases we wanted them to close the gap between varlink and dbus worlds a bit, in that they return a JSON description of the object in question.

Hence, I think we can avoid this all here: let's just return a single json string with the full object from Describe() and that's it? i.e. don't try to half-ass it by passing some parts in an "a{sv}" and then chicken out to add some subobject as json string, but just do the whole thing as json, and thus stay very close to what we did in hostnamed() and networkd()?